File: Problem_1 - 2.py
"""
Time Complexity : O(n). 
Space Complexity : O(n).
    
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.hash = {}
        self.index = None 
        
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not postorder:
            return None 
        self.index = len(postorder) - 1 
    
        for i,j in enumerate(inorder):
            self.hash[j] = i 
        
        def helper(inorer, postorder, s, e):
            if s > e: return None 
            print(self.index)
            root_val = postorder[self.index]
            self.index -= 1 
            
            node = TreeNode(root_val)
            idx = self.hash[root_val]
            
            node.right = helper(inorder, postorder, idx + 1, e)
            node.left = helper(inorder, postorder, s, idx - 1)
            
            return node 
            
        
        return helper(inorder, postorder, 0, len(inorder) - 1)

        
        

File: Problem1_2.java
//Time Complexity: O(n)
//Space Complexity : O(n) + O(h), where h is the height of the tree
//Code run successfully on LeetCode.

public class Problem1_2 {

    HashMap<Integer, Integer> map = new HashMap<>();
    int idx;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        
        if(postorder == null|| postorder.length == 0 || inorder == null|| inorder.length == 0)
            return null;
        
        idx = postorder.length - 1;
        for(int i =0; i < inorder.length; i++)
            map.put(inorder[i], i);
        
        return helper(inorder, postorder, 0, inorder.length -1);
    }
    
    private TreeNode helper(int[] inorder, int[] postorder, int start, int end){
        
        if(start > end)
            return null;
        
        int rootVal = postorder[idx];
        idx--;
        
        TreeNode root = new TreeNode(rootVal);
        int rootIdx = map.get(rootVal);
        
        root.right = helper(inorder, postorder, rootIdx + 1, end);
        root.left = helper(inorder, postorder, start, rootIdx -1);
        
        return root;
    }
}
